Sources/PackageRegistry/RegistryManager.swift

```diff
@@ -0,0 +1,253 @@
+        of package: PackageReference,
+        into fileSystem: FileSystem,
+        at path: AbsolutePath,
+        expectedChecksum: ByteString? = nil,
+        completion: @escaping (Result<Void, Error>) -> Void
+    ) {
+        var components = URLComponents(url: registryBaseURL, resolvingAgainstBaseURL: true)!
+        components.path = "/\(package.identity)" + "/\(version).zip"
+
+        guard let url = components.url else {
+            return completion(.failure(RegistryError.invalidURL))
+        }
+
+        var request = URLRequest(url: url)
+        request.httpMethod = "GET"
+        request.addValue("application/vnd.swift.registry.v1+zip", forHTTPHeaderField: "Accept")
+
+        let task = self.session.downloadTask(with: request) { url, response, error in
+            if let error = error {
+                completion(.failure(error))
+            } else if let response = response as? HTTPURLResponse,
+                      response.statusCode == 200,
+                      response.allHeaderFields["Content-Version"] as? String == "1",
+                      response.allHeaderFields["Content-Type"] as? String == "application/zip",
+                      let digest = response.allHeaderFields["Digest"] as? String,
+                      let url = url,
+                      let data = try? Data(contentsOf: url)
+            {
+                do {
+                    let contents = ByteString(data)
+                    let advertisedChecksum = digest.spm_dropPrefix("sha-256=")
+                    let actualChecksum = SHA256().hash(contents).hexadecimalRepresentation
+
+                    guard (expectedChecksum?.hexadecimalRepresentation ?? actualChecksum) == actualChecksum, advertisedChecksum == actualChecksum else {
+                        throw RegistryError.invalidChecksum(expected: expectedChecksum?.hexadecimalRepresentation ?? advertisedChecksum, actual: actualChecksum)
+                    }
+
+                    try fileSystem.writeFileContents(path, bytes: contents)
+                    completion(.success(()))
+                } catch {
+                    completion(.failure(error))
+                }
+            } else {
+                completion(.failure(RegistryError.invalidResponse))
+            }
+        }
+        task.resume()
+    }
+}
+
+private func loadManifest(
+    _ contents: ByteString,
+    manifestLoader: ManifestLoaderProtocol,
+    toolsVersion: ToolsVersion,
+    swiftLanguageVersion: SwiftLanguageVersion? = nil,
+    packageKind: PackageReference.Kind = .local
+) throws -> Manifest {
+    let fs = InMemoryFileSystem()
+
+    let filename: String
+    if let swiftLanguageVersion = swiftLanguageVersion {
+        filename = Manifest.basename + "@swift-\(swiftLanguageVersion).swift"
+    } else {
+        filename = Manifest.basename + ".swift"
+    }
+
+    try fs.writeFileContents(AbsolutePath.root.appending(component: filename), bytes: contents)
+    return try manifestLoader.load(
```

Comment:
maybe we can expose a function in manifestLoader to load from bytes instead of going through the in-memory FS?

Reply:
I agree. I'll take another look and see if there wasn't a reason why I went with this approach in the first place.

Reply:
Now I remember — `ManifestLoaderProtocol` only supports loading from a filesystem. If we wanted to add support for loading bytes, we'd need to add a new requirement to the protocol, and implement that for all conforming types.  This seemed like a lot of tangential work, but maybe it's worth the extra effort.